core/src/main/java/com/linecorp/armeria/internal/common/stream/AppendingPublisher.java

```diff
@@ -0,0 +1,198 @@
+
+        @SuppressWarnings("rawtypes")
+        private static final AtomicLongFieldUpdater<AppendSubscriber> needToPublishUpdater =
+                AtomicLongFieldUpdater.newUpdater(AppendSubscriber.class, "needToPublish");
+
+        private final Publisher<? extends T> publisher;
+        private final T append;
+        private Subscriber<? super T> downstream;
+        @Nullable
+        private volatile Subscription upstream;
+
+        private volatile long requested;
+        private volatile long needToPublish;
+        private boolean subscribed;
+        private volatile boolean cancelled;
+
+        private boolean appendNeedToSend;
+        private boolean appendSent;
+        private boolean completeSent;
+
+        AppendSubscriber(Publisher<? extends T> publisher, T append, Subscriber<? super T> downstream) {
+            requireNonNull(publisher, "publisher");
+            requireNonNull(append, "append");
+            requireNonNull(downstream, "downstream");
+            this.publisher = publisher;
+            this.append = append;
+            this.downstream = downstream;
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                downstream.onError(new IllegalArgumentException("non-positive request signals are illegal"));
+                return;
+            }
+            if (cancelled || completeSent) {
+                return;
+            }
+            if (appendSent) {
+                complete();
+                return;
+            }
+            if (appendNeedToSend) {
+                append();
+                if (n > 1) {
+                    complete();
+                }
+                return;
+            }
+            for (;;) {
+                final long oldRequested = requested;
+                final long newRequested = LongMath.saturatedAdd(oldRequested, n);
+                if (requestedUpdater.compareAndSet(this, oldRequested, newRequested)) {
+                    if (oldRequested > 0) {
+                        return;
+                    }
+                    break;
+                }
+            }
+            if (!subscribed) {
+                subscribed = true;
+                publisher.subscribe(this);
+            }
+            requestUpstream();
+        }
+
+        private void append() {
+            appendSent = true;
+            downstream.onNext(append);
+        }
+
+        private void complete() {
+            completeSent = true;
+            downstream.onComplete();
+        }
+
+        private void requestUpstream() {
+            for (;;) {
+                final Subscription upstream = this.upstream;
+                final long requested = this.requested;
+                if (upstream == null || requested == 0) {
+                    return;
+                }
+                if (requestedUpdater.compareAndSet(this, requested, 0)) {
+                    needToPublishUpdater.addAndGet(this, requested);
```

Comment:
It seems like this doesn't handle the case when one of the demands is `MAX_VALUE`. Could you handle it as well?

Reply:
✅ updated!